Sephora-M/assignment4/Map.java

```diff
@@ -0,0 +1,109 @@
+package assignment4;
+
+import java.awt.Point;
+import java.util.ArrayList;
+
+
+/**Grid of booleans representing a map. True elements in the grid represent land whereas false tiles represent water.  
+ * 
+ * @author Sephora-M
+ */
+public class Map {
+	private boolean[][] grid;
+	private final int numRows;
+	private final int numCols;
+	
+	/**
+	 * This constructor takes a map of boolean representing the final map.
+	 * @param map, 2d array of boolean, represent the final map
+	 */
+	public Map(boolean[][] map) {
+		if (map == null || map[0].length == 0)
+			throw new IllegalArgumentException("Cannot instatiate an empty or a null grid.");
+		this.grid = map;
+		numCols = map.length;
+		numRows = map[0].length;
+	}
+	
+	/**
+	 * This constructor takes the size of the map and the locations of land tiles.
+	 * @param numRows int, the width of the map
+	 * @param numCols int, the height of the map
+	 */
+	public Map(int numRows, int numCols, ArrayList<Point> landTiles){
+		boolean[][] map = new boolean[numCols][numRows];
+		this.grid = map;
+		this.numCols = numCols;
+		this.numRows = numRows;
+		setLand(landTiles);
+	}
+
+	/**
+	 * Count the number of island in the map.
+	 * @return a int, number of island found
+	 */
+	public int countIslands(){
+		int numIslands = 0;
+		boolean[][] visited = new boolean[numCols][numRows];
+		
+		for(int x = 0; x < numRows; x++) {
+			for(int y = 0; y < numCols; y++) {
+				if (!visited[y][x] && grid[y][x]){
+					visited[y][x] = true;
+					numIslands++;
+					
+					Point firstTileCoord = new Point(x,y);
+					
+					markIsland(firstTileCoord, visited);
+				}
+			}
+		}
+		
+		return numIslands;
+	}
+	
+    private void markIsland(Point startingTile, boolean[][] visited) {
+    	
+    	ArrayList<Point> neighbors = findNeighbors(startingTile, visited);
+    	
+    	for (Point n : neighbors) {
+    		if (getTile(n)){
```

Comment:
let's say we have a grid of all land:
A B
C D
findNeighbors(A) will return {B,C,D} (because none of them have been visited yet). Then you go to B, findNeighbors(B) = {C, D}, then findNeighbors(C) = {D}. When you go back from DFS to B, you visit D for the second time, when you go back to A you also visit C and D again (and call getNeighbors on them). To cut the time in half you can modify this line to:
`if (getTitle(n) & !visited[n.y][n.x]) {`

Reply:
Mmmh.. I actually don't think I'll go to tile B after having done tile A, because in line 51 I do the `!visited[y][x]` check already. Therefore, any island tile that share the A tile will be skipped because they were marked true when searching from A, right?

Reply:
What I meant is that you call findNeighbours first and then call DFS on each of them. It's possible that the n-th neighbour was already visited from a previous neighbour, so there's no need to call markIsland (and findNeighbors) on it again.

You can add a debug message `System.out.println("Getting neighbours for point " + p.x + ", " + p.y)` at the beginning of the findNeighbours method, to see how many times the method is called for the same point (after the point is visited for the first time this method will always return an empty list since all neighbours were already visited, but it slows the algorithm down a bit).

Reply:
Aaah yes I see. Thanks for the much needed further explanation :) fixed!